### PR TITLE
feat: implement CloudEvent down Conversion & Adds Tests

### PR DESCRIPTION
**Pull request by grant**
Refactors event conversion and implements partial event conversion.

- Enables linting on full source code
- Adds many tests to invoker
- Implements CE downcasting
- Refactors CloudEvent conversion
- No external changes to existing functions interface

### Test Output

```
CloudEvents: Utility methods
    ✓ should correctly eval if a CloudEvent is "binary"
    ✓ should correctly eval if a CloudEvent is "structured"
    ✓ should convert requests to a structured CE
    ✓ should get a binary CE from a request

  Invoker: CE -> CE
    ✓ should receive cloudevent from CloudEvents v1.0 structured content mode (39ms)
    ✓ should receive cloudevent from CloudEvents v1.0 binary content mode
    ✓ should receive cloudevent from CloudEvents v1.0 binary content mode with only required fields

  EventConverter: CE -> Legacy
    ✓ should downcast events

  Invoker: CE -> Legacy
    ✓ should receive data and context from CloudEvents v1.0 structured content mode

  EventConverter: Legacy -> CE
    ✓ should upcast Pub/Sub events
    ✓ should upcast legacy Pub/Sub events
    ✓ should upcast Firebase Auth events

  Invoker: HTTP
    ✓ should return the correct HTTP status and call count: empty path
    ✓ should return the correct HTTP status and call count: simple path
    ✓ should return the correct HTTP status and call count: with favicon.ico
    ✓ should return the correct HTTP status and call count: with robots.txt

  Invoker: Legacy Event -> Legacy Event
    ✓ should receive data and context from GCF event
    ✓ should receive data and context from GCF legacy event
    ✓ should receive data and context from GCF event with resource JSON

  Function loader
    ✓ should load function without "main" in package.json
    ✓ should load function with "main" in package.json


  21 passing (107ms)
```

### (Reference) Previous Test Output

```
request to HTTP function
    ✓ should return transformed body: empty path (43ms)
    ✓ should return transformed body: simple path
    ✓ should return transformed body: with favicon.ico
    ✓ should return transformed body: with robots.txt

  GCF event request to event function
    ✓ should receive data and context from GCF event
    ✓ should receive data and context from GCF legacy event
    ✓ should receive data and context from GCF event with resource JSON

  CloudEvents request to event function
    ✓ should receive data and context from CloudEvents v1.0 structured content mode
    ✓ should receive data and context from CloudEvents v1.0 binary content mode

  CloudEvents request to cloudevent function
    ✓ should receive data and context from CloudEvents v1.0 structured content mode
    ✓ should receive data and context from CloudEvents v1.0 binary content mode

  loading function
    ✓ should load function without "main" in package.json
    ✓ should load function with "main" in package.json


  13 passing (99ms)
```

### CODE REVIEW

**Comment by grant** [comment]
Should be ready to review @mtraver. Pardon the large PR.
We can move parts of the changes to a separate PR is needed.

File: src/cloudevents.ts [hunk 1]
```diff
@@ -34,22 +34,89 @@ export function isBinaryCloudEvent(req: express.Request): boolean {
...
+
 /**
  * Returns a CloudEvents context from the given CloudEvents request. Context
  * attributes are retrieved from request headers.
  *
  * @param req Express request object.
  * @return CloudEvents context.
  */
-export function getBinaryCloudEventContext(
+export function getStructuredCloudEventContext(
```
Comment (matthewrobertson) [inline src/cloudevents.ts:106]: I am a bit confused about how this function is different from `convertRequestToStructuredCE`

File: src/invoker.ts [hunk 2]
```diff
@@ -171,16 +172,47 @@ export function wrapCloudEventFunction(
...
+
+        // Clear the property before removing field so the data object
+        // is not deleted.
+        modifiedContext.data = undefined;
+        delete modifiedContext.data;
+        context = modifiedContext;
+      } else {
+        // Set data and context normally
+        data = req.body.data;
+        context = req.body.context;
```
Comment (matthewrobertson) [inline src/invoker.ts:211]: Looks like these variables are already initialized to exactly these values on line 186/187. Can we just remove the else block here or remove the initial values above?

File: src/invoker.ts [hunk 3]
```diff
@@ -171,16 +172,47 @@ export function wrapCloudEventFunction(
...
+
+    // If CloudEvent, convert to legacy event.
+    if (isCloudEvent(req)) {
+      const legacyEvent = convertCloudEventToLegacyEvent(req.headers, req.body);
+      data = legacyEvent.data;
+      context = legacyEvent.context;
+    } else {
+      // If a really legacy event, convert it to a normal event.
+      // See the tests for more details.
+      if (!context) {
```
Comment (matthewrobertson) [inline src/invoker.ts:197]: nit: flatten to `} else if (!context) {`

**Review by matthewrobertson — CHANGES_REQUESTED** [review]
Formally requesting changes because I am sure if you saw my review.

**Comment by grant** [comment]
Thanks. I'll have to apply the comments and likely break this PR into some smaller PRs.

**Comment by grant** [comment]
I've pulled out a few changes into separate PRs. I think it will make things cleaner and safer:

- https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/274
- https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/275
- https://github.com/GoogleCloudPlatform/functions-framework-nodejs/pull/276

**Comment by grant** [comment]
Going to close this PR. Other PRs will be able to complete the conversions. Feel free to ping me if there are questions.

**Pull request closed without merging**